### ISSUE TITLE
Tracking function pointers in indirectly pointed structures

### ISSUE THREAD

**Issue by jprotopopov-ut** [goblint/analyzer#1951]
Goblint revision: e816b4aa7baea44ce2e973b58e39145743eea377
The following snippet:
```c
struct S {
	int type;
	void (*cb)(void);
	struct S *nested;
};

void test(const struct S *s) {
	for (; s && s->type; s++) {
		if (s->type == 1) {
			test(s->nested);
		} else {
			s->cb();
		}
	}
}

void dummy(void) {}

int main() {
	struct S s1[] = {
		{2, dummy},
		{0}
	};
	struct S s2[] = {
		{1, 0, s1},
		{0}
	};

	test(s2);
	return 0;
}
```
causes Goblint to output soundness warning:
```
[Warning][Unsound][Call] No suitable function to be called at call site. Continuing with state before call.
```

The snippet has been manually extracted from `invokeCallbacksPRE_230228` function of [AnalyzeThat rsync](https://gitlab.com/analyzethat/2026/-/raw/main/rsync-v3.4.1.i?ref_type=heads), where the same soundness warning is observed.

**Comment by jprotopopov-ut** [comment]
To confirm, I have just re-tested with adding `static` to the definition of `s1` to make sure that variables are clearly separated spatially, and also recompiled with ubsan. The sample is still functioning as expected.

**Comment by jprotopopov-ut** [comment]
After discussing the issue with @sim642 it appears that the underlying problem is misleading warning and not an unsound analysis in the particular configuration I used. Specifically, due to imprecision of the analysis, Goblint arrives at a context where function pointer can only be null and thus is unable to determine what needs to be called. On other paths, it seems to be aware of the function pointer contents and analyzes the function call properly. The fact that it does not complain about dead code also reinforces this observation. Changing structs domain did not produce improvements. In rsync, callback functions called in this way seem to be analyzed too, at the the ones I have randomly checked.

However, there appears to be an issue with partitioned arrays domain which indeed causes Goblint to conclude that `dummy` function is dead code.
